Make TypeDefault, TypeDerived and VariableType anonymous namespaces

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26865 [wip] Remove manual boxing wrappers
* **#26882 Make TypeDefault, TypeDerived and VariableType anonymous namespaces**
* #26878 Fix binary size in schema inference
* #26869 Fix shared_ptr binary size in op registration
* #26860 Improve binary size of function schema inference
* #26827 Fix c10 registration binary size

Reduce binary size by 500kb by making TypeDerived and VariableType anonymous namespaces instead of classes. TypeDefault is also a namespace now but can't be anonymous because VariableType calls into it.his also has the nice side effect that VariableType.h and ${TypeDerived.h} are much smaller because they don't have to list the operator declarations anymore.

Differential Revision: [D17599686](https://our.internmc.facebook.com/intern/diff/D17599686/)

smessmer committed:
[wip] Make TypeDefault, TypeDerived and VariableType anonymous namespaces



Differential Revision: [D17599686](https://our.internmc.facebook.com/intern/diff/D17599686/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Make TypeDefault, TypeDerived and VariableType anonymous namespaces"


Differential Revision: [D17599686](https://our.internmc.facebook.com/intern/diff/D17599686/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Make TypeDefault, TypeDerived and VariableType anonymous namespaces"


Differential Revision: [D17599686](https://our.internmc.facebook.com/intern/diff/D17599686/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Make TypeDefault, TypeDerived and VariableType anonymous namespaces"


Differential Revision: [D17599686](https://our.internmc.facebook.com/intern/diff/D17599686/)

[ghstack-poisoned]

resistor commented on aten/src/ATen/templates/TypeDefault.cpp:
```diff
@@ -18,11 +18,17 @@
+
 #ifndef USE_STATIC_DISPATCH
+namespace {
 static auto registerer = torch::RegisterOperators()
```
static is no longer needed here

resistor commented on aten/src/ATen/templates/TypeDerived.cpp:
```diff
@@ -42,10 +42,23 @@ Tensor * ${Type}::add(Tensor & a, Tensor & b) {
+
+#ifndef USE_STATIC_DISPATCH
+namespace {
 static auto registerer = torch::RegisterOperators()
```
static is no longer needed here

resistor reviewed (approved):
LGTM with a couple of nits.

smessmer committed:
Update on "Make TypeDefault, TypeDerived and VariableType anonymous namespaces"


Reduce binary size by 500kb by making TypeDerived and VariableType anonymous namespaces instead of classes. TypeDefault is also a namespace now but can't be anonymous because VariableType calls into it.his also has the nice side effect that VariableType.h and ${TypeDerived.h} are much smaller because they don't have to list the operator declarations anymore.

Differential Revision: [D17599686](https://our.internmc.facebook.com/intern/diff/D17599686/)

[ghstack-poisoned]
